README.md
# Get User Custom Field Values

A plugin for WordPress that provides widgets, shortcodes, and template tags to easily retrieve and display custom field values for users.

This plugin can be found in the WordPress Plugin Directory: https://wordpress.org/plugins/get-user-custom-field-values/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Optional: Add filters for `the_user_meta` to filter user custom field data (see the end of the file for commented out samples you may wish to include). And/or add per-meta filters by hooking 'the_user_meta_$field'
4. Give a user a custom field with a value, or have user custom fields already defined. (This generally entails use of plugin(s) that utilize the user custom fields feature built into WordPress. By default, in a practical sense WordPress only sets the `first_name`, `last_name`, and `nickname` user custom fields, so you could try using one of them, even if just for testing. (Though WordPress does provide functions to get those particular fields.)
5. Optional: Go to the Appearance -> Widgets admin page to create one or more 'Get User Custom' sidebar widgets for your widget-enabled theme.
6. Optional: Use the shortcode in a post or page.
7. Optional: Use one of the three template functions provided by this plugin to retrieve the contents of custom fields. You must `echo` the result if you wish to display the value on your site.


## Additional Documentation

* See [readme.txt](https://github.com/coffee2code/get-user-custom-field-values/blob/master/readme.txt) for additional usage information.
* See [DEVELOPER-DOCS.md](DEVELOPER-DOCS.md) for developer-related documentation on hooks, template tags, and shortcode.
* See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/get-user-custom-field-values/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/get-user-custom-field-values/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/get-user-custom-field-values/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 3.3 _(2021-11-09)_

### Highlights:

This recommended release prevents users who can't post unfiltered HTML from using the shortcode in posts (security hardening), adds some new filters, adds DEVELOPER-DOCS.md, notes compatibility through WP 5.8+, and reorganizes and improves unit tests.

### Details:

* Change: Prevent users who cannot post unfiltered HTML from using the shortcode in posts. Props Francesco Carlucci and Erwan.
    * Hardens security to prevent potentail information disclosure or XSS by authors with limited privileges
    * New: Add shortcode class function `can_author_use_shortcodes()`
    * New: Add filter `'get_user_custom_field_values/can_author_use_shortcodes'`
    * Change: Prevent shortcodes created by users who cannot publish unfiltered HTML from being evaulated
    * Change: Prevent display of shortcode builder metabox to users who cannot publish posts
* New: Add filter `'get_user_custom_field_values/show_metabox'` to customize if shortcode builder metabox is shown
* New: Add DEVELOPER-DOCS.md and move template tag and shortcode documentation into it
* Change: Note compatibility through WP 5.8+
* Change: Update copyright date (2021)
* Change: Update shortcode builder widget to 008
* Change: Tweak installation instructions
* Unit tests:
    * Change: Split shortcode-related tests out into their own file
    * Change: Split widget-related tests out into their own file
    * New: Add unit tests for `show_metabox()`
    * New: Add helper function `create_post()`
    * Change: Restructure unit test file structure
        * Change: Move `phpunit/bin/` to `tests/bin/`
        * Change: Move `phpunit/bootstrap.php` into `tests/phpunit/`
        * Change: Move tests from `phpunit/tests/` to `tests/phpunit/tests/`
        * Change: In bootstrap, store path to plugin file constant so its value can be used within that file and in test file
        * Change: In bootstrap, check for test installation in more places and exit with error message if not found
        * Change: Remove 'test-' prefix from unit test files
* New: Add a few more possible TODO items

## 3.2.2 _(2020-09-18)_
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.5+

## 3.2.1 _(2020-06-07)_
* Change: Update shortcode builder widget to 007:
    * New: Store object instantiated during `register()`
    * Change: Cast return value of `c2c_get_user_custom_field_values_post_types` filter as an array
    * Change: Sanitize string used in markup attributes (hardening)
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and added to it)
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Update link to Get Custom Field Values plugin to point to wordpress.org instead of my site
* Change: Unit tests: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and delete commented-out code)

## 3.2 _(2019-12-11)_
* Change: Update widget to 013:
    * Directly load textdomain instead of hooking it to already-fired action
* New: Add README.md
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* Shortcode:
    * Change: Don't show shortcode builder metabox within context of block editor
    * New: Add `show_metabox()`
    * Change: Update version to 006
* Unit tests:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * Change: Tweak whitespace in bootstrap
* Change: Note compatibility through WP 5.3+
* Change: Drop compatibility with version of WP older than 4.6
* Change: Rewrite plugin description
* Change: Update docs to reflect that shortcode builder is not compatible with block editor yet
* Change: Use different markdown formatting for shortcode name to avoid capitalization when displayed in Plugin Directory
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Add GitHub link to readme
* Change: Update copyright date (2020)
* Change: Update License URI to be HTTPS

## 3.1 _(2017-03-14)_
* Fix: Properly handle serialized meta values
* Fix: Properly sanitize field name prior so use as part of a hook name
* Change: Update widget to 012:
    * Correctly pass `$before` and `$after` args in call to `c2c_get_author_custom()`
    * Fix conditional check to properly wrap custom field in span when 'id' and/or 'class' is specified
    * Add `register_widget()` and change to calling it when hooking 'admin_init'
    * Load textdomain
    * Add more substantial unit tests
* Change: Update widget framework:
    * 013:
    * Add `get_config()` as a getter for config array
    * 012:
    * Go back to non-plugin-specific class name of c2c_Widget_012
    * Don't load textdomain
    * Declare class and `load_config()` and `widget_body()` as being abstract
    * Change class variable `$config` from public to protected
    * Discontinue use of `extract()`
    * Apply 'widget_title' filter to widget title
    * Add more inline documentation
    * Minor code reformatting (spacing, bracing, Yoda-ify conditions)
* Change: Update shortcode builder widget to 005:
    * Use `get_config()` to get widget config now that the object variable is protected
    * Add `register()` and change to calling it when hooking 'init'
    * Add more unit tests
* Change: Update unit test bootstrap
    * Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable
    * Enable more error output for unit tests
* Change: Use officially documented order of arguments for `implode()`

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Get_User_Custom_Field_Values
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/get-user-custom-field-values.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Get_User_Custom_Field_Values
 */

define( 'GET_USER_CUSTOM_FIELD_VALUES_PLUGIN_FILE', dirname( __FILE__, 3 ) . '/get-user-custom-field-values.php' );

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

// Backward compatibility (PHPUnit < 6).
$phpunit_backcompat = array(
	'\PHPUnit\Framework\TestCase' => 'PHPUnit_Framework_TestCase',
);
foreach ( $phpunit_backcompat as $new => $old ) {
	if ( ! class_exists( $new ) && class_exists( $old ) ) {
		class_alias( $old, $new );
	}
}

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require GET_USER_CUSTOM_FIELD_VALUES_PLUGIN_FILE;
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/phpunit/tests/widget.php
<?php

defined( 'ABSPATH' ) or die();

class Get_User_Custom_Field_Values_Widget_Test extends WP_UnitTestCase {

	public function tearDown() {
		parent::tearDown();
		$this->unset_current_user();
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function create_user_with_meta( $metas = array(), $user_data = array() ) {
		$user_id = $this->factory->user->create( $user_data );

		if ( empty( $metas ) ) {
			$metas = $this->get_sample_metas();
		}

		foreach ( $metas as $key => $val ) {
			if ( is_array( $val ) ) {
				foreach ( $val as $v ) {
					add_user_meta( $user_id, $key, $v );
				}
			} else {
				add_user_meta( $user_id, $key, $val );
			}
		}

		return $user_id;
	}

	private function widget_init( $config = array() ) {
		$user_id = $this->create_user_with_meta();
		wp_set_current_user( $user_id );

		c2c_GetUserCustomWidget::register_widget();
		$widget = new c2c_GetUserCustomWidget( 'abc_abc', '', array() );

		$default_config = array();
		foreach ( $widget->get_config() as $key => $val ) {
			$default_config[ $key ] = $val['default'];
		}
		$config = array_merge( $default_config, $config );

		if ( true === $config['user_id'] ) {
			$config['user_id'] = $user_id;
		}

		$settings = array( 'before_title' => '', 'after_title' => '', );

		return array( $user_id, $widget, $config, $settings );
	}

	/**
	 * Unsets current user globally. Taken from post.php test.
	 */
	private function unset_current_user() {
		global $current_user, $user_ID;

		$current_user = $user_ID = null;
    }

	private function get_sample_metas() {
		return array(
			'mood'  => 'happy',
			'child' => array( 'adam', 'bob', 'cerise', 'diane' ),
			'color' => array( 'blue', 'white' ),
			'tshirt size' => 'M',
			'location' => 'Denver, CO',
		);
	}


	//
	//
	// TESTS
	//
	//

	/* Widget */

	public function test_widget_class_exists() {
		$this->assertTrue( class_exists( 'c2c_GetUserCustomWidget' ) );
	}

	public function test_widget_version() {
		$this->assertEquals( '013', c2c_GetUserCustomWidget::version() );
	}

	public function test_widget_framework_class_name() {
		$this->assertTrue( class_exists( 'c2c_Widget_013' ) );
	}

	public function test_widget_framework_version() {
		$this->assertEquals( '013', c2c_Widget_013::version() );
	}

	public function test_widget_hooks_widgets_init() {
		$this->assertEquals( 10, has_filter( 'widgets_init', array( 'c2c_GetUserCustomWidget', 'register_widget' ) ) );
	}

	public function test_widget_made_available() {
		$this->assertArrayHasKey( 'c2c_GetUserCustomWidget', $GLOBALS['wp_widget_factory']->widgets );
	}

	public function test_widget_body_with_class() {
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'class' => 'abcd', 'user_id' => true )  );

		$this->assertEquals( '<span class="abcd">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_id() {
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'id' => 'myid', 'user_id' => true ) );

		$this->assertEquals( '<span id="myid">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_class_and_id() {
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'class' => 'abcd', 'id' => 'myid', 'user_id' => true ) );

		$this->assertEquals( '<span id="myid" class="abcd">happy</span>', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_body_with_class_and_id_but_no_meta_value() {
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'nonexistent', 'class' => 'abcd', 'id' => 'myid', 'user_id' => true ) );

		$this->assertEmpty( $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_user_id_of_current_in_invalid_situation() {
		$author_id = $this->create_user_with_meta( array( 'mood' => 'confused' ) );

		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'user_id' => 'current' ) );

		$this->assertNull( $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_user_id_of_current() {
		$author_id = $this->create_user_with_meta( array( 'mood' => 'perplexed' ) );
		$user_id = $this->create_user_with_meta();
		wp_set_current_user( $user_id );

		// Simulate conditions when it is valid to run.
		$GLOBALS['authordata'] = get_userdata( $author_id );
		$GLOBALS['wp_query']->is_single = true;

		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'user_id' => 'current' ) );

		$this->assertEquals( 'perplexed', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_explicit_user_id() {
		$u_id = $this->create_user_with_meta( array( 'mood' => 'joyous' ) );
		$user_id = $this->create_user_with_meta();

		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood', 'user_id' => $u_id ) );

		$this->assertEquals( 'joyous', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_no_user_id() {
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood' ) );

		$u_id = $this->create_user_with_meta( array( 'mood' => 'jazzed' ) );
		wp_set_current_user( $u_id );


		$this->assertEquals( 'jazzed', $widget->widget_body( $config, '', $settings ) );
	}

	public function test_widget_with_no_user_id_and_no_current_user() {
		$user_id = $this->create_user_with_meta( array( 'mood' => 'befuddled' ) );
		list( $user_id, $widget, $config, $settings ) = $this->widget_init( array( 'field' => 'mood' ) );

		$this->unset_current_user();

		$this->assertNull( $widget->widget_body( $config, '', $settings ) );
	}

}


get-user-custom.widget.php
<?php
/**
 * Get User Custom Field Values plugin widget code.
 *
 * Copyright (c) 2004-2021 by Scott Reilly (aka coffee2code)
 *
 * @package c2c_GetUserCustomWidget
 * @author  Scott Reilly
 * @version 013
 */

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_GetUserCustomWidget' ) ) :

require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'c2c-widget.php' );

class c2c_GetUserCustomWidget extends c2c_Widget_013 {

	/**
	 * Returns version of the widget.
	 *
	 * @since 009
	 *
	 * @return string
	 */
	public static function version() {
		return '013';
	}

	/**
	 * Registers the widget.
	 *
	 * @since 012
	 */
	public static function register_widget() {
		register_widget( __CLASS__ );
	}

	/**
	 * Constructor
	 */
	public function __construct() {
		parent::__construct( 'get-user-custom', __FILE__, array( 'width' => 300 ) );

		$this->load_textdomain();

		add_filter( $this->get_hook( 'excluded_form_options' ), array( $this, 'excluded_form_options' ) );
	}

	/**
	 * Loads the plugin textdomain.
	 *
	 * @since 012
	 */
	public function load_textdomain() {
		load_plugin_textdomain( 'get-user-custom-field-values' );
	}

	/**
	 * Initializes the plugin's configuration and localizable text variables.
	 */
	public function load_config() {
		$this->title       = __( 'Get User Custom Field', 'get-user-custom-field-values' );
		$this->description = __( 'A list of custom field value(s) from user(s).', 'get-user-custom-field-values' );

		$this->config = array(
			// input can be 'checkbox', 'multiselect', 'select', 'short_text', 'text', 'textarea', 'hidden', or 'none'
			// datatype can be 'array' or 'hash'
			// can also specify input_attributes
			'title' => array(
				'input'   => 'text',
				'default' => __( 'User Custom Field', 'get-user-custom-field-values' ),
				'label'   => __( 'Title', 'get-user-custom-field-values' ),
			),
			'field' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'Custom field key', 'get-user-custom-field-values' ),
				'help'    => __( '<strong>*Required.</strong>  The name of the custom field key whose value you wish to have displayed.', 'get-user-custom-field-values' ),
			),
			'this_post' => array(
				'input'   => 'checkbox',
				'default' => false,
				'label'   => __( 'This post\'s author?', 'get-user-custom-field-values' ),
				'help'    => __( 'For the author of the post containing this shortcode. Takes precedence over \'User ID\'', 'get-user-custom-field-values' ),
			),
			'user_id' => array(
				'input'   => 'short_text',
				'default' => '',
				'label'   => __( 'User ID', 'get-user-custom-field-values' ),
				'help'    => __( 'ID of user whose custom field\'s value you want to display. Leave blank to search for the custom field for the currently logged in user. Use <code>0</code> to indicate it should only work on the permalink page for a page/post.', 'get-user-custom-field-values' ),
			),
			'before' => array(
				'input' => 'text',
				'default' => '',
				'label'   => __( 'Before text', 'get-user-custom-field-values' ),
				'help'    => __( 'Text to display before the custom field.', 'get-user-custom-field-values' ),
			),
			'after' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'After text', 'get-user-custom-field-values' ),
				'help'    => __( 'Text to display after the custom field.', 'get-user-custom-field-values' ),
			),
			'none' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'None text', 'get-user-custom-field-values' ),
				'help'    => __( 'Text to display if no matching custom field is found (or it has no value). Leave this blank if you don\'t want anything to display when no match is found.', 'get-user-custom-field-values' ),
			),
			'between' => array(
				'input'   => 'text',
				'default' => ', ',
				'label'   => __( 'Between text', 'get-user-custom-field-values' ),
				'help'    => __( 'Text to display between custom field items if more than one are being shown.', 'get-user-custom-field-values' ),
			),
			'before_last' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'Before last text', 'get-user-custom-field-values' ),
				'help'    => __( 'Text to display between the second to last and last custom field items if more than one are being shown.', 'get-user-custom-field-values' ),
			),
			'id' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'HTML id', 'get-user-custom-field-values' ),
				'help'    => __( 'The \'id\' attribute for the &lt;span&gt; tag to surrounds output.', 'get-user-custom-field-values' ),
			),
			'class' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'HTML class', 'get-user-custom-field-values' ),
				'help'    => __( 'The \'class\' attribute for the &lt;span&gt; tag to surrounds output.', 'get-user-custom-field-values' ),
			),
		);
	}

	/**
	 * Outputs the body of the widget.
	 *
	 * @param array $args Widget args.
	 * @param array $instance Widget instance.
	 * @param array $settings Widget settings.
	 */
	public function widget_body( $args, $instance, $settings ) {
		extract( $args );
		extract( $settings );

		$ret = '';

		// Determine, based on inputs given, which template tag to use.
		if ( 0 === $user_id || $this_post ) {
			$user_id = 'current';
		}

		if ( $user_id ) {
			if ( 'current' == $user_id ) {
				$ret = c2c_get_author_custom( $field, $before, $after, $none, $between, $before_last );
			} else {
				$ret = c2c_get_user_custom( $user_id, $field, $before, $after, $none, $between, $before_last );
			}
		} else {
			$ret = c2c_get_current_user_custom( $field, $before, $after, $none, $between, $before_last );
		}

		// If either 'id' or 'class' attribute was defined, then wrap output in span
		if ( $ret && ( $id || $class ) ) {
			$tag = '<span';

			if ( $id ) {
				$tag .= ' id="' . esc_attr( $id ) . '"';
			}

			if ( $class ) {
				$tag .= ' class="' . esc_attr( $class ) . '"';
			}

			$tag .= ">$ret</span>";
			$ret = $tag;
		}

		return $ret;
	}

	/**
	 * Validates widget instance values.
	 *
	 * @param array  $instance Widget instance values.
	 * @return array The filtered widget instance values.
	 */
	public function validate( $instance ) {
		$instance['field']   = trim( $instance['field'] );
		$instance['user_id'] = trim( $instance['user_id'] );
		if ( $instance['user_id'] ) {
			$instance['user_id'] = intval( $instance['user_id'] );
		}

		return $instance;
	}

	/**
	 * Defines widget form options that shouldn't be shown by default (since they are used for the shortcode widget).
	 *
	 * @param array  $excluded_form_options Array of form options that shouldn't be shown.
	 * @return array The array of form options that shouldn't be shown.
	 */
	public function excluded_form_options( $excluded_form_options ) {
		if ( $excluded_form_options === null ) {
			$excluded_form_options = array( 'this_post' );
		}

		return $excluded_form_options;
	}

} // end class c2c_GetUserCustomWidget

add_action( 'widgets_init', array( 'c2c_GetUserCustomWidget', 'register_widget' ) );

endif; // end if !class_exists()


get-user-custom.shortcode.php
<?php
/**
 * Get User Custom Field Values plugin shortcode code.
 *
 * Copyright (c) 2004-2021 by Scott Reilly (aka coffee2code)
 *
 * @package c2c_GetUserCustomFieldValuesShortcode
 * @author  Scott Reilly
 * @version 008
 */

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_GetUserCustomFieldValuesShortcode' ) && class_exists( 'c2c_GetUserCustomWidget' ) ) :

class c2c_GetUserCustomFieldValuesShortcode {
	/**
	 * The instance of this object.
	 *
	 * @access public
	 * @since 007
	 * @var c2c_GetUserCustomFieldValuesShortcode
	 */
	public static $instance;

	public $name           = 'shortcode_get_user_custom_field_values';
	public $shortcode      = 'user_custom_field';
	public $title          = '';
	public $widget_handler = '';
	public $widget_base    = '';

	/**
	 * Returns version of the class.
	 *
	 * @since 003
	 *
	 * @return string
	 */
	public static function version() {
		return '008';
	}

	/**
	 * Registers the shortcode.
	 *
	 * @since 005
	 */
	public static function register() {
		self::$instance = new self( $GLOBALS['wp_widget_factory']->widgets['c2c_GetUserCustomWidget'] );
	}

	/**
	 * Constructor.
	 */
	public function __construct( $widget_handler ) {
		$this->title          = __( 'Get User Custom Field Values - Shortcode', 'get-user-custom-field-values' );
		$this->widget_handler = $widget_handler;
		$this->widget_base    = 'widget-' . $this->widget_handler->id_base;
		$this->shortcode      = apply_filters( 'c2c_get_user_custom_field_values_shortcode', $this->shortcode );

		add_shortcode( $this->shortcode,         array( $this, 'shortcode' ) );
		add_action( 'load-post.php',             array( $this, 'register_post_page_hooks' ) );
		add_action( 'load-post-new.php',         array( $this, 'register_post_page_hooks' ) );
		add_filter( 'default_hidden_meta_boxes', array( $this, 'default_hidden_meta_boxes' ), 10, 2 );
	}

	/**
	 * Filters/actions to hook on the admin post.php page.
	 *
	 * @since 2.8
	 */
	public function register_post_page_hooks() {
		add_action( 'admin_print_footer_scripts', array( $this, 'admin_js' ) );
		add_action( 'do_meta_boxes',              array( $this, 'do_meta_box' ), 10, 3 );
	}

	/**
	 * Determines if post author can use shortcodes.
	 *
	 * @since 3.3
	 *
	 * @param null|int|WP_User $user The user ID, object, or null to refer to
	 *                               current post's author. Default null.
	 * @param null|int|WP_Post $post The post ID, object, or null to refer to
	 *                               current post. Default null.
	 * @return bool True if post author can use shortcode, else false.
	 */
	public function can_author_use_shortcodes( $user = null, $post = null ) {
		// Default to allowing shortcode usage as shortcode may appear in non-post contexts.
		// It is assumed those other contexts are trusted.
		$can = true;

		// If post not provided, use global post.
		$post = get_post( $post );

		// If user provided, get its object. Else, assume current post's author.
		if ( $user ) {
			if ( is_int( $user ) ) {
				$user = get_userdata( (int) $user );
			}
		} else {
			if ( $post ) {
				$user = get_userdata( $post->post_author );
			}
		}

		// User must have 'unfiltered_html' capability, which requires a user to
		// be known.
		if ( $user instanceof WP_User ) {
			$can = user_can( $user->ID, 'unfiltered_html' );
		}

		/**
		 * Filters if a given author (if applicable given the shortcode context),
		 * can use the 'user_custom_field' shortcode, or to have their use of it
		 * evaluated.
		 *
		 * Note: In contexts where no author is relevant (such as a shortcode
		 * appearing in a sidebar widget), then use of the shortcode is
		 * assumed to be allowed and trusted.
		 *
		 * @since 3.3
		 *
		 * @param bool          $can  Is shortcode use permitted?
		 * @param WP_User|false $user User object, or false if no user in context.
		 * @param WP_Post|false $post Post object, or false if no post in context.
		 */
		return (bool) apply_filters( 'get_user_custom_field_values/can_author_use_shortcodes', $can, $user, $post );
	}

	/**
	 * Determines if the metabox should be shown.
	 *
	 * The metabox should only be shown in the classic editor.
	 *
	 * @since 006
	 * @since 3.3 Prevent users who cannot publish posts from seeing the metabox. Add `'get_user_custom_field_values/show_metabox'` filter.
	 *
	 * @return bool True if the metabox can be shown, false otherwise.
	 */
	public function show_metabox() {
		$current_screen = get_current_screen();
		$show = true;

		// Ensure current screen is for a post.
		if ( $show ) {
			$show = 'post' === $current_screen->base;
		}

		// Ensure block editor is not enabled (metabox is not compatible in that context).
		if ( $show ) {
			$show = ! method_exists( $current_screen, 'is_block_editor' ) || ! $current_screen->is_block_editor();
		}

		// Ensure post author can use shortcodes.
		if ( $show ) {
			/**
			 * Filters if the shortcode builder metabox is shown.
			 *
			 * @ssince 3.3
			 *
			 * @param bool $show Show the metabox? Default is whether the post author can publish unfiltered HTML.
			 */
			$show = (bool) apply_filters( 'get_user_custom_field_values/show_metabox', $this->can_author_use_shortcodes() );
		}

		return $show;
	}

	/**
	 * Register meta box.
	 *
	 * By default, the shortcode builder is present for all post types. Filter
	 * 'c2c_get_custom_field_values_post_types' to limit its use.
	 *
	 * @since 3.4
	 *
	 * @param string  $post_type The post type
	 * @param string  $type      The mode for the meta box (normal, advanced, or side)
	 * @param WP_Post $post      The post
	 */
	public function do_meta_box( $post_type, $type, $post ) {
		if ( ! $this->show_metabox() ) {
			return;
		}

		$post_types = (array) apply_filters( 'c2c_get_user_custom_field_values_post_types', get_post_types() );
		if ( ! in_array( $post_type, $post_types ) ) {
			return;
		}

		add_meta_box( $this->name, $this->title, array( $this, 'form' ), $post_type, 'side', 'core' );
	}

	/**
	 * Hides the shortcode wizard by default.
	 *
	 * @since 2.8
	 *
	 * @param array     $hidden List of all hidden metaboxes
	 * @param WP_Screen $screen Screen object.
	 */
	public function default_hidden_meta_boxes( $hidden, $screen ) {
		if ( ! in_array( $this->name, $hidden ) ) {
			$hidden[] = $this->name;
		}

		return $hidden;
	}

	/**
	 * Shortcode handler.
	 *
	 * @since 3.3 Prevent evaluation of shortcodes in posts author by users who cannot publish unfiltered HTML.
	 *
	 * @param array  $atts    The shortcode attributes
	 * @param string $content The text wrapped by the shortcode's opening and closing tags.
	 * @return string
	 */
	public function shortcode( $atts, $content = null ) {
		$defaults = array();

		if ( ! $this->can_author_use_shortcodes() ) {
			return '';
		}

		foreach ( $this->widget_handler->get_config() as $opt => $values ) {
			$defaults[ $opt ] = isset( $values['default'] ) ? $values['default'] : '';
		}

		$atts2 = shortcode_atts( $defaults, $atts );

		foreach ( array_keys( $this->widget_handler->get_config() ) as $key ) {
			$$key = $atts2[ $key ];
		}

		$ret = '';

		if ( '0' === $user_id || $this_post ) {
			$user_id = 'current';
		}

		if ( $user_id ) {
			if ( 'current' == $user_id ) {
				$ret = c2c_get_author_custom( $field, $before, $after, $none, $between, $before_last );
			} else {
				$ret = c2c_get_user_custom( $user_id, $field, $before, $after, $none, $between, $before_last );
			}
		} else {
			$ret = c2c_get_current_user_custom( $field, $before, $after, $none, $between, $before_last );
		}

		// If either 'id' or 'class' attribute was defined, then wrap output in span
		if ( $ret && ( $id || $class ) ) {
			$tag = '<span';
			if ( $id ) {
				$tag .= ' id="' . esc_attr( $id ) . '"';
			}
			if ( $class ) {
				$tag .= ' class="' . esc_attr( $class ) . '"';
			}
			$tag .= ">$ret</span>";
			$ret = $tag;
		}

		return $ret;
	}

	/**
	 * Outputs the JS for the shortcode wizard.
	 */
	public function admin_js() {
		if ( ! $this->show_metabox() ) {
			return;
		}

		echo <<<JS
		<script type="text/javascript">
			var {$this->name} = function () {}

			{$this->name}.prototype = {
				options           : {},
				generateShortCode : function() {
					var content = this['options']['content'];
					delete this['options']['content'];

					var attrs = '';
					jQuery.each(this['options'], function(name, value){
						if (value != '') {
							attrs += ' ' + name + '="' + value + '"';
						}
					});
					return '[{$this->shortcode}' + attrs + ' /]';
				},
				sendToEditor      : function(f) {
					var collection = jQuery(f).find("input[id^={$this->widget_base}]:not(input:checkbox), \
										input[id^={$this->widget_base}]:checkbox:checked, \
										textarea[id^={$this->widget_base}]");
					var \$this = this;
					collection.each(function () {
						var name = this.name.substring(this.name.lastIndexOf('[')+1, this.name.length-1);
						if (\$this['options'][name] == undefined)
							\$this['options'][name] = this.value;
						else
							\$this['options'][name] += ', '+this.value;
					});
					send_to_editor(this.generateShortCode());
					/* Delete data after generating shortcode so that the form can be used to generate another shortcode */
					collection.each(function () {
						var name = this.name.substring(this.name.lastIndexOf('[')+1, this.name.length-1);
						delete \$this['options'][name];
					});
					return false;
				}
			}

			var admin_{$this->name} = new {$this->name}();
		</script>

JS;
	}

	/**
	 * Outputs shortcode wizard form.
	 */
	public function form() {
		$this->widget_handler->form( array(), array( 'title' ) );
		printf(
			'<p class="submit"><input type="button" class="button-primary" onclick="return admin_%s.sendToEditor(this.form);" value="%s" /></p>' . "\n",
			esc_attr( $this->name ),
			esc_attr__( 'Send shortcode to editor', 'get-user-custom-field-values' )
		);
	}

} // end class c2c_GetUserCustomFieldValuesShortcode

add_action( 'init', array( 'c2c_GetUserCustomFieldValuesShortcode', 'register' ), 11 );

endif; // end if !class_exists()
